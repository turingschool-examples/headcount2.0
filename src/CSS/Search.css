.search-bar {
  margin: 1rem 0.5rem;
}

.search-input {
  height: 2rem;
  width: 20rem;
  border: 1.5px solid #ccc;
  margin: 1rem 0;
  border-radius: 5px;
}

.search-input:hover {
  box-shadow: 1px 1px 1px 1px #ccc;
}

.hide-suggestions {
  display: none;
}

.did-you-mean {
  font-style: italic;
  color: gray;
}

.show-suggestions {
  margin: 1rem 0;
  text-align: center;
}

.suggestions {
  cursor: pointer;
  border: 1px solid gray;
  margin: 0;
  padding: 2px;
  text-align: center;
}

.suggestions:hover {
  background-color: rgb(244, 243, 243);
  box-shadow: 1.5px 1.5px 1.5px 1.5px #ccc;
}

button {
  width: 21.3rem;
  height: 2rem;
  border-radius: 5px;
  background-color: rgb(230, 230, 230);
  font-size: 1.1rem;
}

button:hover {
  background-color: rgb(142, 142, 142);
  color: rgb(237, 237, 237);
  cursor: pointer;
}

input[type="text"] {
  font-size: 1.1rem;
  line-height: 5rem;
  padding-left: 1rem;
}

.show {
  margin-bottom: 10rem;
  font-style: italic;
}

.show-none {
  display: none;
}

.display-error {
  padding: 1rem 0;
  text-align: center;
}
