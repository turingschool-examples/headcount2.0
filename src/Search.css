.search {
  text-align: center
}

.inputField {
  padding: 10px;
  border: 3px solid red;
  /*margin: 10px;*/
  width: 300px;
}

.inputButton {
  padding: 10px;
  background-color: white;
  margin: 10px;
  border-radius: 30%;
  border: 3px solid green;
  color: green;
}

.inputButton:hover {
  background-color: green;
  color: white;
}
