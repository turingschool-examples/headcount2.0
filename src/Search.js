import React, { Component } from "react";
import DistrictRepository from "./helper";
import kindergarners from "./data/kindergartners_in_full_day_program";

export default class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userInput: "",
      searchSuggestions: []
    };
    this.districtRepository = new DistrictRepository(kindergarners);
  }

  handleChange = ({ name, value }) => {
    this.setState({ [name]: value });
    this.suggestDistricts(this.state.userInput);
  };

  suggestDistricts = e => {
    let suggestions = this.districtRepository.findAllMatches(
      this.state.userInput
    );
    this.setState({ searchSuggestions: suggestions });
    if (e.length > 2) {
      this.props.handleSubmit(this.state.searchSuggestions);
    }
  };

  searchDistrict = district => {
    const filteredDistricts = this.districtRepository.findByName(district);
    this.props.handleSubmit([filteredDistricts]);
  };

  render() {
    return (
      <div>
        <form>
          <input
            placeholder="Search District"
            name="userInput"
            onChange={e => this.handleChange(e.target)}
            value={this.state.userInput}
          />
          {/* <button onClick={e => this.searchDistrict(e)}>Search</button> */}
        </form>
        <div>
          {this.state.searchSuggestions.map((district, i) => {
            if (
              this.state.searchSuggestions &&
              this.state.userInput.length > 3
            ) {
              return (
                <p
                  className="suggestions"
                  key={i}
                  onClick={e => this.searchDistrict(e.target.textContent)}
                >
                  {district.location}
                </p>
              );
            }
          })}
        </div>
      </div>
    );
  }
}
