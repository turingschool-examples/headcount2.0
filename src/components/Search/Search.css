/*SEARCH COMPONENT RELATED CSS GOES HERE*/
.Search {
  width: 35%;
  display: flex;
  justify-content: space-between
}

input {
  height: 2rem;
  width: 100%;
  border: 2px solid #BEBEBE;
  padding-left: .6rem;
  font-size: 1.25rem;
  letter-spacing: 0.5px;
  color: #006989;
}

::placeholder {
  font-family: "Roboto", sans-serif;
  font-weight: 400;
  color: #9E9E9E;
}