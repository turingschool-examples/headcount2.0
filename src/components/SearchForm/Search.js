import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class Search extends Component {
  constructor() {
    super();
    this.state = {
      query: ''
    };
  }

  handleChange =  event => {
    const { value } = event.target;

    this.props.filterDistricts(value);

    this.setState({ query: value });
  }

  render() {
    const { query } = this.state;

    return (
      <input
        onChange={this.handleChange}
        type="text"
        name="query"
        value={query}
        placeholder="Search"
      />
    );
  }
}

Search.propTypes = {
  filterDistricts: PropTypes.func.isRequired
};

export default Search;
