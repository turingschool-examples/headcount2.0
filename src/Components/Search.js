import React, { Component } from "react";
import DistrictRepository from "../helper";
import kindergarners from "../data/kindergartners_in_full_day_program";
import PropTypes from "prop-types";
import "../CSS/Search.css";

export default class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      districtInputOne: "",
      searchSuggestions: []
    };
    this.districtRepository = new DistrictRepository(kindergarners);
  }

  handleChange = ({ value }) => {
    this.setState({ districtInputOne: value });
    this.suggestDistricts();
  };

  suggestDistricts = () => {
    const { districtInputOne } = this.state;
    const searchSuggestions = this.districtRepository.findAllMatches(
      districtInputOne
    );
    this.props.selectedDistricts.forEach(district => {
      searchSuggestions.push(district);
    });

    this.setState({ searchSuggestions });
    if (districtInputOne.length > 1) {
      this.props.handleSubmit(searchSuggestions);
    } else {
      const newDistricts = this.districtRepository.districtsArray();
      this.props.handleSubmit(newDistricts);
    }
  };

  clearInput = () => {
    this.setState({ districtInputOne: "" });
    const newDistricts = this.districtRepository.districtsArray();
    this.props.handleSubmit(newDistricts);
  };

  searchDistrict = district => {
    const filteredDistricts = this.districtRepository.findByName(district);
    this.props.handleSubmit([filteredDistricts]);
  };

  render() {
    return (
      <div className="search-bar">
        <form>
          <input
            type="text"
            className="search-input"
            placeholder={
              this.props.selectedDistricts.length > 0
                ? "Remove Card to Search"
                : "Search District"
            }
            onChange={event => this.handleChange(event.target)}
            value={this.state.districtInputOne}
            disabled={this.props.selectedDistricts.length > 1}
          />
          <button onClick={() => this.props.clearComparisons()}>
            Clear Fields
          </button>
        </form>
        <div
          className={
            this.state.districtInputOne.length > 1
              ? "show-suggestions"
              : "hide-suggestions"
          }
        >
          <span
            className={
              this.state.searchSuggestions.length
                ? "show-none"
                : "display-error"
            }
          >
            None Found! Search Again
          </span>

          <span
            className={
              this.state.searchSuggestions.length ? "show" : "show-none"
            }
          >
            Did you Mean?
          </span>
          {this.state.searchSuggestions.map((district, index) => {
            if (
              this.state.searchSuggestions &&
              this.state.districtInputOne.length > 1
            ) {
              var suggestions = (
                <p
                  className="suggestions"
                  key={index}
                  onClick={event => {
                    this.props.selectCard(event.target.textContent);
                    this.clearInput();
                  }}
                >
                  {district.location}
                </p>
              );
            }
            return suggestions;
          })}
        </div>
      </div>
    );
  }
}

const { func, arrayOf, shape, bool, string, number } = PropTypes;

Search.propTypes = {
  selectedDistricts: arrayOf(
    shape({
      clicked: bool,
      location: string,
      stats: shape({ year: number })
    })
  ),
  handleSubmit: func.isRequired,
  selectCard: func.isRequired,
  clearComparisons: func.isRequired
};
