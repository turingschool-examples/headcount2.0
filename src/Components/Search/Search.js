import React from 'react';
import PropTypes from 'prop-types';
import './Search.css';

const Search = ( {handleSearch} ) => {
  return (
      <div className="search-container">
        <input 
        placeholder="Search for district..."
        className="search"
        onChange={ (e) => handleSearch(e.target.value) }/>
      </div>
    )
}

Search.propTypes = {
  handleSearch: PropTypes.func.isRequired
}

export default Search;