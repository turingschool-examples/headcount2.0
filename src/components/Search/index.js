import React, { Component } from 'react'
import PropTypes from 'prop-types'
import './Search.css'

class Search extends Component {
  constructor({findAllMatches}) {
    super()
    this.state = {
      query: ''
    }
  }

  handleInputChange = event => {
    const { value } = event.target

    this.setState({query: value})
    this.props.findAllMatches(value);
  }

  handleSubmit = event => {
    event.preventDefault()
  }

  render() {
    return(
      <form className="Search" onSubmit={this.handleSubmit} >
        <input
          className="search-input"
          aria-label="Search for a specific district here"
          placeholder="Search for a specific district"
          name="query"
          value={this.state.query}
          onChange={this.handleInputChange}
        />
      </form>
    )
  }
}

Search.propTypes = {
  findAllMatches: PropTypes.func.isRequired
}

export default Search