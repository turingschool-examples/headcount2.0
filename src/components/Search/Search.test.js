import React from 'react'
import { shallow, mount } from 'enzyme'
import Search from './index'

describe('Search', () => {
  let findAllMatches;
  let handleInputMock;
  let wrapper;

  beforeEach(() => {
    findAllMatches = jest.fn()
    handleInputMock = jest.fn()
    wrapper = shallow(<Search findAllMatches={findAllMatches}/>)
  })

  it('should have a search input', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should call handleInputChange when the input changes', () => {
    wrapper = mount(<Search findAllMatches={findAllMatches}/>)
    const spy = spyOn(wrapper.instance(), 'handleInputChange')
    wrapper.instance().forceUpdate()
    const mockEvent = { target: { value: 'something' } }

    wrapper.find('.search-input').simulate('change', mockEvent)

    expect(spy).toHaveBeenCalled()
  })

  it('updates state when handleInputChange is called', () => {
    const mockEvent = { target: { name: 'query', value: 'something' } }

    wrapper.instance().handleInputChange(mockEvent)

    expect(wrapper.state('query')).toBe('something')
  })

  it('calls findAllMatches when handleInputChange is called', () => {
    const mockEvent = { target: { name: 'query', value: 'something' } }

    wrapper.instance().handleInputChange(mockEvent)

    expect(findAllMatches).toHaveBeenCalled()
  })

  it('should call handleSubmit onSubmit of the form', () => {
    wrapper = mount(<Search findAllMatches={findAllMatches}/>)
    const spy = spyOn(wrapper.instance(), 'handleSubmit')
    const mockEvent = { preventDefault: jest.fn() }

    wrapper.instance().forceUpdate()

    wrapper.find('form').simulate('submit', mockEvent)

    expect(spy).toHaveBeenCalled()
  })
})