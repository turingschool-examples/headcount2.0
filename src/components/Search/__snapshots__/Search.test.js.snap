// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should have a search input 1`] = `
<form
  className="Search"
  onSubmit={[Function]}
>
  <input
    aria-label="Search for a specific district here"
    className="search-input"
    name="query"
    onChange={[Function]}
    placeholder="Search for a specific district"
    value=""
  />
</form>
`;
